// fitnode/snsauth.js
var snsauth = function (db, config) {
    this.db = db;
    this.config = config;
    //console.log(config.roles);
}

exports.snsauth = snsauth;

snsauth.prototype = {
    constructor: snsauth,
    ids: function (role_code, account, rel_account, strId, callback) {
        var auth = this;
        this.db.collection('user_relation').find({
            //account: account, rel_account: rel_account, //以后加
            "roles.code": role_code
        }).toArray(function (err, items) {
            var w = [];
            if (items.length == 1)
                for (var i in items[0].roles) {
                    var role = items[0].roles[i];

                    var funcs = auth.config.sysroles[role.code].funcs;
                    for (var j in funcs) {
                        var func = auth.config.sysfuncs[funcs[j]];
                        if (func.doc)
                            if (strId) w.push(role.doc_id);
                            else w.push(auth.db.ObjectID(role.doc_id));
                    }
                }
            callback(w);
        });
    },
    check: function (id, role_code, account, rel_account, callback) {
        ids(role_code, account, rel_account, true, function (w) {
            if (w.indexOf(id) >= 0) return true;
            else return false;
        });
    },
    limit: function (query, field, role_code, account, rel_account, callback) {
        var limitId = (!field || field == "_id");
        this.ids(role_code, account, rel_account, limitId ? false : true, function (w) {
            if (limitId)
                query["_id"] = { "$in": w };
            else query[field] = { "$in": w };
            console.log(query);
            callback(query);
        });
    },
    filter: function (query, field, role_code, account, rel_account, callback) {
        this.ids(role_code, account, rel_account, true, function (w) {
            var result = [];
            query.each(function (err, item) {
                if (!item) callback(result)
                else {
                    if (w.indexOf(item._id.toString()) >= 0) result.push(item);
                }
            });
        });
    },
    demo_data: function () {
        this.db.collection('user_define_roles').remove().insert({
            account: '4f61a6b86e435c09b885ad0a', name: '基友', sysroles: ['friend', 'dev']
        });

        this.db.collection('user_relation').remove().insert({
            account: '4f61a6b86e435c09b885ad0a',
            rel_account: '4f619a9365631f0d14dc3b2c', //lobby@163.com
            roles: [
                { code: 'friend', doc_id: null },
                { code: 'dev', doc_id: '4f619a9365631f0d14dc3b2d' },
                { code: 'dev', doc_id: '4f61852865631f03146075c8' },
            //{ id: 'xxx', doc_id: null }
            ]
        });
    }
};
