// fitnode/util.js
/**********  顺序执行n个函数 ***********/
var aaa = function () {
    return this;
}
exports.aaa= aaa;

var seq_async = function (funcs, callback) {
    this.funcs = funcs;
    this.callback = callback;
}

exports.seq_async = seq_async;

seq_async.prototype = {
    constructor: seq_async,
    execOne: function (params, index) {
        var obj = this;
        this.funcs[index](params, function (data) {
            if (arguments.length <= 1) data = data;
            else data = arguments;

            if (index == obj.funcs.length - 1)
                obj.callback(data);
            else obj.execOne(data, index + 1);
        })
    },
    exec: function () {
        this.execOne(null, 0);
    }
}

/**********  只执行一个函数，执行次数由数组决定     ***********/
exports.seq_asyncArray = function (func, paramsArray, callback) {
    this.func = func;
    this.paramsArray = paramsArray;
    this.callback = callback;
}


exports.seq_asyncArray.prototype = {
    constructor: exports.seq_asyncArray,
    execOne: function (params, index) {
        var obj = this;
        this.func(this.paramsArray[index], params, function (data) {
            if (arguments.length <= 1) data = data;
            else data = arguments;

            if (index == obj.paramsArray.length - 1)
                obj.callback(data);
            else obj.execOne(data, index + 1);
        })
    },
    exec: function () {
        if (this.paramsArray && this.paramsArray.length > 0) this.execOne(null, 0);
        else this.callback(null);
    }
}
